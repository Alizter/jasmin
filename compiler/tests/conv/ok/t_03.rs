fn bar1(x: reg! (b64),y: reg! (b64), c: reg! (b1)) -> (reg! (b64), reg! (b1), reg! (b64)) {
    code! {
        x = sub_v(x,y);
    }
    return (x,c,y)
}

fn foo1(x: reg! (b64), y: reg! (b64), c: reg! (b1)) -> (reg! (b64),reg! (b1),reg! (b64)) {
    code! {
        x = add_v(x,y);
        (x,c,y) = inl!{ bar1(x,y,c) };
    }
    return (x,c,y)
}

fn test() {
    var! {
        u:   reg!    (b64);
        v:   reg!    (b64);
        cf:  reg!    (b1);
        u_:  reg!    (b64);
        v_:  reg!    (b64);
        cf_: reg!    (b1);
        i:   inline! (b64);
    }

    code! {
        u = 0;
        v = 0;

        u_,cf_,v_ = inl!{ foo1(u,v,cf) };
        for i in (0..10) {
            while cf {
                if cf {
                    u_,cf_,v_ = inl!{ foo1(u,v,cf) };
                } else {
                    u_,cf_,v_ = inl! { bar1(u,v,cf) };
                }
            }
        }
    }
}

/*
START:CMD
#ARG="renumber_module_unique,save[/tmp/before.mil][],cert_inline[test],save[/tmp/after.mil][]"
ARG="renumber_module_unique,cert_inline[test],print[after_inlining]"
END:CMD
*/
