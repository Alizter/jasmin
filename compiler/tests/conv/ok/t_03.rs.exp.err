Processed file tests/conv/ok/t_03.rs
