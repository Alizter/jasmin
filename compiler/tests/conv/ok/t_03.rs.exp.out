renumbering identifiers in all function(s)
   REMOVED_TIME
inlining all calls (certified) in ``test'' function(s)
   REMOVED_TIME
>> after_inlining:
#[macro_use] extern crate jasmin;

fn bar1(mut x.1: reg! (b64), mut y.3: reg! (b64), mut c.2: reg! (b1)) -> (reg! (b64),reg! (b1),reg! (b64)) {
                                       code! {
                                           x.1 = sub_v(x.1,y.3);
                                       }
                                       return (x.1,c.2,y.3)
                                   }

fn foo1(mut x.4: reg! (b64), mut y.6: reg! (b64), mut c.5: reg! (b1)) -> (reg! (b64),reg! (b1),reg! (b64)) {
    var! {
        c.2: reg! (b1)
        x.1: reg! (b64)
        y.3: reg! (b64);
    }
    code! {
        x.4 = add_v(x.4,y.6);
        x.1 := x.4;
        y.3 := y.6;
        c.2 := c.5;
        x.1 = sub_v(x.1,y.3);
        x.4 := x.1;
        c.5 := c.2;
        y.6 := y.3;
    }
    return (x.4,c.5,y.6)
}

fn test() {
    var! {
        c.2: reg! (b1)
        c.5: reg! (b1)
        cf.9: reg! (b1)
        cf_.11: reg! (b1)
        i.13: inline (b64)
        u.7: reg! (b64)
        u_.10: reg! (b64)
        v.8: reg! (b64)
        v_.12: reg! (b64)
        x.1: reg! (b64)
        x.4: reg! (b64)
        y.3: reg! (b64)
        y.6: reg! (b64);
    }
    code! {
        u.7 = jc!(0);
        v.8 = jc!(0);
        x.4 := u.7;
        y.6 := v.8;
        c.5 := cf.9;
        x.4 = add_v(x.4,y.6);
        x.1 := x.4;
        y.3 := y.6;
        c.2 := c.5;
        x.1 = sub_v(x.1,y.3);
        x.4 := x.1;
        c.5 := c.2;
        y.6 := y.3;
        u_.10 := x.4;
        cf_.11 := c.5;
        v_.12 := y.6;
        for i.13 in jc!(0)..jc!(10) {
            while cf.9 {
                if cf.9 {
                    x.4 := u.7;
                    y.6 := v.8;
                    c.5 := cf.9;
                    x.4 = add_v(x.4,y.6);
                    x.1 := x.4;
                    y.3 := y.6;
                    c.2 := c.5;
                    x.1 = sub_v(x.1,y.3);
                    x.4 := x.1;
                    c.5 := c.2;
                    y.6 := y.3;
                    u_.10 := x.4;
                    cf_.11 := c.5;
                    v_.12 := y.6;
                } else {
                  x.1 := u.7;
                  y.3 := v.8;
                  c.2 := cf.9;
                  x.1 = sub_v(x.1,y.3);
                  u_.10 := x.1;
                  cf_.11 := c.2;
                  v_.12 := y.3;
                }
            }
        }
    }
}


total transformation time: REMOVED_TIME

total time (with save/print): REMOVED_TIME
